Content.Server/Magic/MagicSystem.cs

```diff
@@ -0,0 +1,312 @@
+    {
+        if (args.Handled)
+            return;
+
+        //Get the position of the player
+        var transform = Transform(args.Performer);
+        var coords = transform.Coordinates;
+
+        SoundSystem.Play(Filter.Pvs(coords), args.KnockSound.GetSound(), AudioParams.Default.WithVolume(args.KnockVolume));
+
+        //Look for doors and don't open them if they're already open.
+        foreach (var entity in _lookup.GetEntitiesInRange(coords, args.Range))
+        {
+            if (TryComp<AirlockComponent>(entity, out var airlock))
+                airlock.BoltsDown = false;
+
+            if (TryComp<DoorComponent>(entity, out var doorComp) && doorComp.State is not DoorState.Open)
+                _doorSystem.StartOpening(doorComp.Owner);
+        }
+
+        args.Handled = true;
+    }
+
+    /// <summary>
+    /// Spawns entity prototypes from a list within range of click.
+    /// </summary>
+    /// <remarks>
+    /// It will offset mobs after the first mob based on the OffsetVector2 property supplied.
+    /// </remarks>
+    /// <param name="args"> The Spawn Spell Event args.</param>
+    private void OnSpawnSpell(SpawnSpellEvent args)
+    {
+        if (args.Handled)
+            return;
+
+        var targetMapCoords = args.Target;
+
+        SpawnSpellHelper(args.Contents, targetMapCoords, args.TemporarySummon, args.OffsetVector2, args.Lifetime, _random);
+
+        args.Handled = true;
+    }
+
+    /// <summary>
+    /// Loops through a supplied list of entity prototypes and spawns them
+    /// </summary>
+    /// <remarks>
+    /// If an offset of 0, 0 is supplied then the entities will all spawn on the same tile.
+    /// Any other offset will spawn entities starting from the source Map Coordinates and will increment the supplied
+    /// offset
+    /// </remarks>
+    /// <param name="entityEntries"> The list of Entities to spawn in</param>
+    /// <param name="mapCoords"> Map Coordinates where the entities will spawn</param>
+    /// <param name="temporarySummon"> Check to see if the entities should self delete</param>
+    /// <param name="offsetVector2"> A Vector2 offset that the entities will spawn in</param>
+    /// <param name="lifetime"> Sets the <see cref="SpawnSpellComponent"/> lifetime from the <see cref="SpawnSpellEvent"/> args</param>
+    /// <param name="random"> Resolves param, check out <see cref="EntitySpawnEntry"/> for what to put on the prototype</param>
+    private void SpawnSpellHelper(List<EntitySpawnEntry> entityEntries, MapCoordinates mapCoords, bool temporarySummon, Vector2 offsetVector2, float lifetime, IRobustRandom? random)
+    {
+        var getProtos = EntitySpawnCollection.GetSpawns(entityEntries, random);
+
+        var offsetCoords = mapCoords;
+        foreach (var proto in getProtos)
+        {
+            var entity = Spawn(proto, offsetCoords);
+            entity.SnapToGrid();
+            offsetCoords = offsetCoords.Offset(offsetVector2);
```

Comment:
Should probably check the entity isn't already anchored before snapping.

Reply:
How does this work? Are certain entities spawned, spawn anchored and does that mess with the snapping or something?